send on closed channel in the TestRandomFail

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

```
 [2024/03/25 08:11:41.578 +00:00] [ERROR] [agg_util.go:55] ["parallel hash aggregation panicked"] [error="Intest panic: final worker is panicked before start"] [errorVerbose="Intest panic: final worker is panicked before start\ngithub.com/pingcap/tidb/pkg/util.GetRecoverError\n\tpkg/util/util.go:305\ngithub.com/pingcap/tidb/pkg/executor/aggregate.recoveryHashAgg\n\tpkg/executor/aggregate/agg_util.go:53\ngithub.com/pingcap/tidb/pkg/executor/aggregate.(*HashAggFinalWorker).cleanup\n\tpkg/executor/aggregate/agg_hash_final_worker.go:252\nruntime.gopanic\n\tGOROOT/src/runtime/panic.go:920\ngithub.com/pingcap/tidb/pkg/executor/aggregate.intestBeforeFinalWorkerStart\n\tpkg/executor/aggregate/agg_hash_final_worker.go:263\ngithub.com/pingcap/tidb/pkg/executor/aggregate.(*HashAggFinalWorker).run\n\tpkg/executor/aggregate/agg_hash_final_worker.go:234\nruntime.goexit\n\tsrc/runtime/asm_amd64.s:1650"] [stack="github.com/pingcap/tidb/pkg/executor/aggregate.recoveryHashAgg\n\tpkg/executor/aggregate/agg_util.go:55\ngithub.com/pingcap/tidb/pkg/executor/aggregate.(*HashAggFinalWorker).cleanup\n\tpkg/executor/aggregate/agg_hash_final_worker.go:252\nruntime.gopanic\n\tGOROOT/src/runtime/panic.go:920\ngithub.com/pingcap/tidb/pkg/executor/aggregate.intestBeforeFinalWorkerStart\n\tpkg/executor/aggregate/agg_hash_final_worker.go:263\ngithub.com/pingcap/tidb/pkg/executor/aggregate.(*HashAggFinalWorker).run\n\tpkg/executor/aggregate/agg_hash_final_worker.go:234"] [stack="github.com/pingcap/tidb/pkg/executor/aggregate.recoveryHashAgg\n\tpkg/executor/aggregate/agg_util.go:55\ngithub.com/pingcap/tidb/pkg/executor/aggregate.(*HashAggFinalWorker).cleanup\n\tpkg/executor/aggregate/agg_hash_final_worker.go:252\nruntime.gopanic\n\tGOROOT/src/runtime/panic.go:920\ngithub.com/pingcap/tidb/pkg/executor/aggregate.intestBeforeFinalWorkerStart\n\tpkg/executor/aggregate/agg_hash_final_worker.go:263\ngithub.com/pingcap/tidb/pkg/executor/aggregate.(*HashAggFinalWorker).run\n\tpkg/executor/aggregate/agg_hash_final_worker.go:234"]
panic: send on closed channel [recovered]
	panic: send on closed channel
goroutine 601 [running]:
github.com/pingcap/tidb/pkg/executor/aggregate.recoveryHashAgg(0xc00087db58?, {0x47b2c40, 0x6601a20})
	pkg/executor/aggregate/agg_util.go:54 +0x110
github.com/pingcap/tidb/pkg/executor/aggregate.(*HashAggExec).fetchChildData.func1()
	pkg/executor/aggregate/agg_hash_executor.go:434 +0x86
panic({0x47b2c40?, 0x6601a20?})
	GOROOT/src/runtime/panic.go:920 +0x270
github.com/pingcap/tidb/pkg/executor/aggregate.(*HashAggExec).fetchChildData(0xc000b6a000, {0x66334f0, 0x9d664a0}, 0xc0054625f0)
	pkg/executor/aggregate/agg_hash_executor.go:481 +0x4fe
created by github.com/pingcap/tidb/pkg/executor/aggregate.(*HashAggExec).prepare4ParallelExec in goroutine 90
	pkg/executor/aggregate/agg_hash_executor.go:545 +0x19a 
```
### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

--- a/pkg/executor/aggregate/agg_hash_executor.go
+++ b/pkg/executor/aggregate/agg_hash_executor.go
@@ -184,7 +184,7 @@ func (e *HashAggExec) Close() error {
 	}
 	if e.parallelExecValid {
 		// `Close` may be called after `Open` without calling `Next` in test.
-		if !e.prepared.Load() {
+		if e.prepared.CompareAndSwap(false, true) {
 			close(e.inputCh)
 			for _, ch := range e.partialOutputChs {
 				close(ch)

